Add  Dynamic Output Shape Tag For data-dependent ops, handle in FakeTensor

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79170
* #78972
* #78523
* #78522
* #78677

Adds a tag for operators which have an output whose shape depends on input Tensor data. This tag is tested through the fake_tensor tests. If an operator is not tagged appropriately, the fake tensor test will fallback to cpu to run the operator, and then output metadata will diverge. This helped me find `linalg_lstsq` among others. 

I included in the tags CompositeImplicitAutograd ops, although I'm not sure what the general policy is for these operators. 

One difficulty wrt/fake tensors was [aten::one_hot](https://codebrowser.bddppq.com/pytorch/pytorch/aten/src/ATen/native/Onehot.cpp.html#23). Because it is a `CompositeImplicitAutograd`, it won't be seen by `__torch_dispatch__`. It would be possible to handle this case either 
a) always throwing with `local_scalar_dense`, but this would add a huge amount of false negatives where we think operators have data-dependent output shapes but are actually not. 
b) Maybe I should also be defining `__torch_function__`  for fake_tensor to error on specific composites. (this didn't work when I tried it, didn't get invoked, so maybe there's something I'm missing)


Feel free to suggest a different name than `dynamic_output_shape`.

eellison committed:
Add  Dynamic Output Shape Tagdfor ata-dependent ops, handle in FakeTensor

[ghstack-poisoned]

eellison committed:
Update on "Add  Dynamic Output Shape Tagdfor ata-dependent ops, handle in FakeTensor"




[ghstack-poisoned]

eellison committed:
Update on "Add  Dynamic Output Shape Tag For data-dependent ops, handle in FakeTensor"




[ghstack-poisoned]

ezyang commented on test/test_ops.py:
```diff
@@ -1525,16 +1540,28 @@ def map_to_fake(e):
             except torch._subclasses.fake_tensor.SparseInputException:
                 pass
+            except torch._subclasses.fake_tensor.DynamicOutputShapeException:
+                self.assertTrue(name in dynamic_output_op_tests)
```
do you want a `not in` assert when this succeeds?

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -21,6 +23,33 @@ class ComplexInputException(Exception):
+
+
+# TODO: use tags when available
+# operators whose output shape depends on input tensor data
```
cc @anjali411

eellison committed:
Update on "Add  Dynamic Output Shape Tag For data-dependent ops, handle in FakeTensor"


Adds a tag for operators which have an output whose shape depends on input Tensor data. This tag is tested through the fake_tensor tests. If an operator is not tagged appropriately, the fake tensor test will fallback to cpu to run the operator, and then output metadata will diverge. This helped me find `linalg_lstsq` among others. 

I included in the tags CompositeImplicitAutograd ops, although I'm not sure what the general policy is for these operators. 

One difficulty wrt/fake tensors was [aten::one_hot](https://codebrowser.bddppq.com/pytorch/pytorch/aten/src/ATen/native/Onehot.cpp.html#23). Because it is a `CompositeImplicitAutograd`, it won't be seen by `__torch_dispatch__`. It would be possible to handle this case either 
a) always throwing with `local_scalar_dense`, but this would add a huge amount of false negatives where we think operators have data-dependent output shapes but are actually not. 
b) Maybe I should also be defining `__torch_function__`  for fake_tensor to error on specific composites. (this didn't work when I tried it, didn't get invoked, so maybe there's something I'm missing)


Feel free to suggest a different name than `dynamic_output_shape`.


[ghstack-poisoned]

eellison commented:
@pytorchbot merge
